core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/Stream.scala

```diff
@@ -20,38 +16,80 @@ abstract class Stream[+T] { self =>
 
   private[this] var listeners = Vector.empty[T => Unit]
 
+  private var hasShutdown = false
+  def subscribeToParents(): Unit
+  def unsubscribeFromParents(): Unit
+  def checkRelaunch(): Unit = {}
+
   protected def publishValue(value: T @uncheckedVariance): Unit = {
     listeners.foreach(_.apply(value))
     // TODO: more stuff maybe
   }
 
+  /**
+    * Adds a listener for elements of this stream. Callbacks will be executed
+    * whenever a new value is published in order of when the callbacks were added.
+    * Callbacks added first will be called first and callbacks added last will
+    * be called last.
+    *
+    * @param thunk the listener to be called on each published value
+    * @return a function to call to remove the listener
+    */
+  def foreach(thunk: T => Unit): Cancel = {
+    self.synchronized {
+      listeners = listeners :+ thunk
+      if (listeners.size == 1) {
+        // first listener, boot up!
+        if (hasShutdown) {
+          checkRelaunch()
```

Comment:
Don't you need to set hasShutDown to false here?

Reply:
That only tracks if the stream has booted up at some point, not the current state of the stream.